### DISCUSSION TITLE
Is it possible to use Gitoxide without a file system?

### DISCUSSION THREAD

**Discussion by rimutaka** [GitoxideLabs/gitoxide#1150] [Q&A] [answered]
I am thinking of building a browser extension (WASM/Rust) for repo analysis, but there seems to be no file access API.
I'd have to clone the repo into memory, somehow and work from there.

Is it possible with Gitoxide?

**Comment by Byron** [comment, marked as answer]
No, that's not possible yet.

There is [a tracking-ticket](https://github.com/Byron/gitoxide/issues/463) that you might want to follow for details and updates.

**Comment by rimutaka** [comment]
Thanks @Byron! I'm happy to help with the WASM stream if you think it's worth putting effort into at this stage.

**Reply by Byron** [reply]
I think that depends on *how* this could be achieved. My suggestion is to look at `gix-pack` and see how that could be compiled with WASI, what it would take and if that would even work for you. Maybe before writing any code, you could share your analysis in the tracking ticket so I can get a feel for what it means. Thanks for your consideration.

**Reply by rimutaka** [reply]
@Byron , I tried to modify gix-pack for wasm32-unknown-unknown and ran into some fundamental blockers. I wonder if we should address them first.

#### WASM target

My interest is to make it work in the browser, probably as a service worker (a long running b/g process, primarily used by browser extensions). That means `wasm32-unknown-unknown` target with all the IO and system level calls limitations.

#### rust-analyzer crashes if the default target is `wasm32-unknown-unknown`

`cargo check` works for `wasm32-unknown-unknown`, but setting it as the default target for rust-analyzer in VSCode crashes the RA server.
I would track down and disable the dependencies that have cmake/clang steps. Not sure how far it will get us because they are at the core of everything.

#### Lack of local storage

We'll need some kind of persistent storage, eventually. The only cross-browser option that I know of is IndexedDB ([MDN](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API), [web-sys crate bindings](https://rustwasm.github.io/wasm-bindgen/api/web_sys/struct.IdbDatabase.html). It's not the easiest API to work with. Not sure how realistic this option is.

**Reply by Byron** [reply]
Thanks for trying and for sharing your results. Making this work is going to take major effort, and I think it's something that might happen after `gix` gets stabilized in a year or two. Once it has all the features and is used more, that would be one of the major capabilities left to add.

> rust-analyzer crashes if the default target is `wasm32-unknown-unknown`

At least for this one there should be way to fix it easily as there is the `max-pure` configuration option for `gitoxide`, which boils down to a configuration that will avoid any/most C-bindings in dependencies. `cmake` is definitely not required anymore then.
